### DISCUSSION TITLE
Contract.from_abi not working as expected

### DISCUSSION THREAD

**Discussion by AlbertoSinigaglia** [smartcontractkit/full-blockchain-solidity-course-py#35] [Show and tell]
in the video (at arout 7:22:42) you are using this code to load a contract from an ABI:
```python
Contract.from_abi(contract._name, contract_addres, contract._abi)
```
however this is not working if we run the integration test (it says that there is no attribute called `_abi` and `_name` as we can see from [here](https://eth-brownie.readthedocs.io/en/stable/api-network.html#contractcontainer))... I've solved using this:
```python
contract = Contract.from_abi(contract_type.__class__.__name__, contract_addr, contract_type.abi)
#                                         ^^^^^^^^^^^^^^^^--- don't do this, use ._name as @PatrickAlphaC answered
```

also, for me 60 seconds to wait wasn't enough... but 4 minutes yes

**Comment by PatrickAlphaC** [comment]
Hmm.... This "works" but not in the way that we'd expect. Not a big deal... but let's figure out what's going on. 

`contract_type` should be `<class 'brownie.network.contract.ContractContainer'>`
If you do `print(type(contract_type))` right before that line and run the deploy script, what do you get? 

`contract_type.__class__.__name__` will be `ContractContainer`
but we want `LinkToken`, which we get from `contract_type._name`

Could you check what's going on?

**Reply by AlbertoSinigaglia** [reply]
ok yes I can confirm that `._name` is working fine, thank you so much

**Comment by Niferu** [comment]
Hello, I was searching whole brownie documentation for `ContractContainer`, but I couldn't find attribute `_name` for it, we did not pre defined it and there is no such argument in any of mock we are deploying. Can anyone explain to me where such argument is located and how we were supposed to know calling `contract_type._name` will return it's name?

If it is from below, why there is no attribute `.abi`?

![image](https://user-images.githubusercontent.com/110537615/188766845-0d0cae96-ad74-4f3f-af6c-0a6f16a896a5.png)

**Reply by cromewar** [reply]
@PatrickAlphaC a hand on this one please.

**Reply by PatrickAlphaC** [reply]
```
$ brownie console
>>> from brownie import APIConsumer
>>> APIConsumer.__dir__()
["tx", "bytecode", "_contracts", "_project", "_build", "_sources", "topics", "selectors", "signatures", "deploy", "_flattener", "__module__", "__doc__", "__init__", "__iter__", "__getitem__", "__delitem__", "__len__", "__repr__", "_reset", "_revert", "remove", "at", "_add_from_tx", "get_verification_info", "publish_source", "_slice_source", "_dir_color", "abi", "_name", "info", "get_method", "decode_input", "__dict__", "__weakref__", "__new__", "__hash__", "__str__", "__getattribute__", "__setattr__", "__delattr__", "__lt__", "__le__", "__eq__", "__ne__", "__gt__", "__ge__", "__reduce_ex__", "__reduce__", "__subclasshook__", "__init_subclass__", "__format__", "__sizeof__", "__dir__", "__class__"]
```

The `__dir__()` function in python shows all the functions and attributes you can call on an object and you can see the `_name` in there. The documentation likely just doesn't have it in there, but if you look in the source code you can see that attribute is exposed. 

And, there is a `.abi` attribute.

**Reply by Niferu** [reply]
@PatrickAlphaC Thanks a lot for that! As I'm new to all of this, that `__dir__()` function will help me a lot with clarifying things.
